proof/invariant-abstract/Syscall_AI.thy

```diff
@@ -1321,18 +1321,21 @@ lemma update_time_stamp_invs[wp]:
   "update_time_stamp \<lbrace>invs\<rbrace>"
   by (wpsimp simp: update_time_stamp_def)
 
-crunches update_time_stamp
-  for scheduler_action[wp]: "\<lambda>s. P (scheduler_action s)"
-  and cur_thread[wp]: "\<lambda>s. P (cur_thread s)"
-  and ct_in_state[wp]: "ct_in_state P"
-  and sc_at_pred_n[wp]: "\<lambda>s. Q (sc_at_pred_n N proj P p s)"
-  and pred_tcb_at[wp]: "\<lambda>s. Q (pred_tcb_at p P t s)"
-  and pred_tcb_at_ct[wp]: "\<lambda>s. pred_tcb_at p P (cur_thread s) s"
+lemma
+  shows preemption_point_valid_objs[wp]: "preemption_point \<lbrace>valid_objs\<rbrace>"
+  and preemption_point_pspace_aligned[wp]: "preemption_point \<lbrace>pspace_aligned\<rbrace>"
+  and preemption_point_pspace_distinct[wp]: "preemption_point \<lbrace>pspace_distinct\<rbrace>"
+  and preemption_point_pspace_no_overlap[wp]: "preemption_point \<lbrace>pspace_no_overlap S\<rbrace>"
+  and preemption_point_cte_wp_at[wp]: "preemption_point \<lbrace>cte_wp_at P p\<rbrace>"
+  and preemption_point_cur_sc_tcb[wp]: "preemption_point \<lbrace>cur_sc_tcb\<rbrace>"
+  and preemption_point_invs[wp]: "preemption_point \<lbrace>invs\<rbrace>"
+  by (wpsimp wp: preemption_point_inv simp: cur_sc_tcb_def)+
```

Comment:
I think I wanted all of these lemmas for some of the work in Refine. If you try to crunch these, it'll say that there's no theorem proved, because it says `preemption_point_inv` will work. But I wanted these facts available as is in Refine, because using `preemption_point_inv` seriously messed things up. I think some of those proofs were extremely tricky and I didn't want to make them any worse. 

Does that make sense, or should there be a better of going about that?

Reply:
You might be able to crunch them with `preemption_point_inv` passed in as a `rule` parameter instead of `wp`. From memory, `crunch` looks at the `wp` set for theorems that could be used to prove the goal, but the `rule` set passed in is used differently.

Reply:
Ah yes that worked, thanks!

Reply:
Awesome. Let me know if you want an explanation about how `rule` parameters are actually used by `crunch`.